falcon_kit/mains/pr_ctg_track.py

```diff
@@ -0,0 +1,157 @@
+
+Reader = io.CapturedProcessReaderContext
+
+
+def get_pid_to_ctg(fn):
+    pid_to_ctg = {}
+    with open(fn) as f:
+        for row in f:
+            row = row.strip().split()
+            pid, rid, oid, ctg = row
+            pid_to_ctg.setdefault( pid, set()  )
+            pid_to_ctg[ pid ].add(ctg)
+    return pid_to_ctg
+
+def run_tr_stage1(db_fn, fn, min_len, bestn, pid_to_ctg):
+    cmd = "LA4Falcon -mo %s %s" % (db_fn, fn)
+    reader = Reader(cmd)
+    with reader:
+        return fn, tr_stage1(reader.readlines, min_len, bestn, pid_to_ctg)
+
+def tr_stage1(readlines, min_len, bestn, pid_to_ctg):
+    """
+    for each read in the b-read column inside the LAS files, we
+    keep top `bestn` hits with a priority queue through all overlaps
+    """
+    rtn = {}
+    for l in readlines():
+        l = l.strip().split()
+        q_id, t_id = l[:2]
+        overlap_len = -int(l[2])
+        idt = float(l[3])
+        q_s, q_e, q_l = int(l[5]), int(l[6]), int(l[7])
+        t_s, t_e, t_l = int(l[9]), int(l[10]), int(l[11])
+        if t_l < min_len:
+            continue
+        if q_id not in pid_to_ctg:
+            continue
+        rtn.setdefault(t_id, [])
+        if len(rtn[t_id]) < bestn:
+            heappush(rtn[t_id], (overlap_len, q_id) )
+        else:
+            heappushpop(rtn[t_id], (overlap_len, q_id) )
+
+    return rtn
+
+def run_track_reads(exe_pool, base_dir, file_list, min_len, bestn, db_fn):
+    io.LOG('preparing tr_stage1')
+    io.logstats()
+    asm_dir = os.path.abspath( os.path.join(base_dir, "2-asm-falcon") )
+    pid_to_ctg = get_pid_to_ctg( os.path.join(asm_dir, "read_maps/read_to_contig_map" ) )
+    inputs = []
+    for fn in file_list:
+        if len(fn) != 0:
+            inputs.append( (run_tr_stage1, db_fn, fn, min_len, bestn, pid_to_ctg) )
+
+    """
+    Aggregate hits from each individual LAS and keep the best n hit. 
+    Note that this does not guarantee that the final results is globally the best n hits espcially 
+    when the number of `bestn` is too small.  In those case, if there is more hits from single LAS
+    file, then we will miss some good  hits.
+    """
+
+    bread_to_areads = {}
+    for fn, res in exe_pool.imap(io.run_func, inputs):
+        for k in res:
+            bread_to_areads.setdefault(k, [])
+            for item in res[k]:
+                if len(bread_to_areads[k]) < bestn:
+                    heappush( bread_to_areads[k], item )
+                else:
+                    heappushpop( bread_to_areads[k], item )
+
+
+    with open( os.path.join(asm_dir, "read_maps/pread_to_contigs"), "w") as out_f:
+        for bread in bread_to_areads:
+
+            ctg_score = {}
+            for s, pid in bread_to_areads[bread]:
+                if pid not in pid_to_ctg: continue
+                
+                ctgs = pid_to_ctg[pid]
+                for ctg in ctgs:
+                    ctg_score.setdefault(ctg, [0,0])
+                    ctg_score[ctg][0] += -s
```

Comment:
I'm kidding. Machs nichts to me, and your way might be more clear. I'd never thought about it before.